### PR TITLE
GEO out of keyspace

### PR DESCRIPTION
**Pull request by ashtul**

### CODE REVIEW

**Comment by MeirShpilraien** [comment]
@ashtul did you see that there is a compilation error?

**Comment by ashtul** [comment]
@MeirShpilraien it works fine on my system and compiles on circleCI

**Comment by MeirShpilraien** [comment]
@ashtul looks like everything is failing

**Comment by ashtul** [comment]
@meir this is built over the index_out_of_key_space branch which has failures.

**Comment by ashtul** [comment]
@MeirShpilraien  this is built over  `index_out_of_key_space` which had some failures before.
Have other branches that fix some of these issues. Or should I try and merge to master?

**Comment by MeirShpilraien** [comment]
@ashtul I do not think that on index_out_of_key_space all the tests failed (or did they?)
Maybe we need to merge master into it so it will run with Redis 6 ?

**Comment by ashtul** [comment]
@MeirShpilraien many failed before. On my machine, nothing directly related to geo fails.
Made other PRs for these tests.

**Comment by MeirShpilraien** [comment]
@ashtul on my machine everything passed expect for geo tests I think ..

File: src/document.c [hunk 1]
```diff
@@ -463,19 +460,19 @@ FIELD_PREPROCESSOR(geoPreprocessor) {
...
-  fdata->geoSlon = c;
-  fdata->geoSlat = pos;
-  return 0;
-}
 
-FIELD_BULK_INDEXER(geoIndexer) {
-  GeoIndex gi = {.ctx = ctx, .sp = fs};
-  int rv = GeoIndex_AddStrings(&gi, aCtx->doc.docId, fdata->geoSlon, fdata->geoSlat);
+  char *end1 = NULL, *end2 = NULL;
+  double lon = strtod(c, &end1);
```
Comment (MeirShpilraien) [inline src/document.c:465]: Can this fail?

Comment (MeirShpilraien) [reply src/document.c:465]: Ohh you check it next line :+1:

File: src/geo_index.c [hunk 2]
```diff
@@ -83,6 +54,13 @@ int GeoFilter_Parse(GeoFilter *gf, ArgsCursor *ac, QueryError *status) {
 
 void GeoFilter_Free(GeoFilter *gf) {
   if (gf->property) rm_free((char *)gf->property);
+  if (gf->numericFilters) {
+    for (int i = 0; i < GEO_RANGE_COUNT; ++i) {
+      if (gf->numericFilters[i])
+        NumericFilter_Free(gf->numericFilters[i]);
+    }
+  }
+  rm_free(gf->numericFilters);
```
Comment (MeirShpilraien) [inline src/geo_index.c:63]: Shouldn't this be inside the if?

File: src/geo_index.c [hunk 3]
```diff
@@ -125,16 +103,31 @@ static t_docId *geoRangeLoad(const GeoIndex *gi, const GeoFilter *gf, size_t *nu
...
+              NewNumericFilter(ranges[ii].min, ranges[ii].max, 1, 1);
+      filt->fieldName = rm_strdup(gf->property);
+      filt->geoFilter = gf;
+      struct indexIterator *numIter = NewNumericFilterIterator(ctx, filt, NULL, INDEXFLD_T_GEO);
+      if (numIter != NULL) {
+        iters[itersCount++] = numIter;
+      }
+    }
+  }
+  IndexIterator *it = NewUnionIterator(iters, itersCount, NULL, 1, 1);
```
Comment (MeirShpilraien) [inline src/geo_index.c:126]: If itersCount is one there is no need to wrap it in a Union iterator. Also if its zero you can just return NULL no?

File: src/numeric_index.c [hunk 4]
```diff
@@ -336,7 +336,8 @@ IndexIterator *NewNumericRangeIterator(const IndexSpec *sp, NumericRange *nr,
                                        const NumericFilter *f) {
 
   // if this range is at either end of the filter, we need to check each record
-  if (NumericFilter_Match(f, nr->minVal) && NumericFilter_Match(f, nr->maxVal)) {
+  if (NumericFilter_Match(f, nr->minVal) && NumericFilter_Match(f, nr->maxVal) &&
+      f->geoFilter == NULL) {
```
Comment (MeirShpilraien) [inline src/numeric_index.c:340]: Why is this needed? if MIN and MAX are not matched but we do have goeFilter we still have not results no?

File: src/pytest/test.py [hunk 5]
```diff
@@ -918,9 +918,9 @@ def testGeoErrors(env):
     env.expect('ft.add idx hotel 1.0 fields name hill location -0.1757,51.5156').equal('OK')
     env.expect('ft.search idx hilton geofilter location -0.1757 51.5156 1 km').equal([0L])
 
-    # Insert error
-    env.expect('ft.add', 'idx', 'hotel1', 1, 'fields', 'name', '_hotel1', 'location', '1, 1').error()   \
-            .contains('Could not index geo value')
+    # Insert error - works fine with out of keyspace implementation
```
Comment (MeirShpilraien) [inline src/pytest/test.py:921]: Why is it working fine?

File: src/pytest/test.py [hunk 6]
```diff
@@ -934,13 +934,6 @@ def testGeoErrors(env):
     env.expect('ft.search idx hilton geofilter location -0.1757 51.5156 1').error()   \
             .contains('GEOFILTER requires 5 arguments')
 
-    if not env.isCluster():
```
Comment (MeirShpilraien) [inline src/pytest/test.py:937]: :+1:

File: src/pytest/test.py [hunk 7]
```diff
@@ -961,24 +954,20 @@ def testGeo(env):
...
 
         res = gsearch('hilton', "-0.1757", "51.5156", '10')
         env.assertEqual(14, res[0])
-        env.assertEqual('hotel93', res[1])
-        env.assertEqual('hotel92', res[3])
-        env.assertEqual('hotel79', res[5])
 
         res2 = gsearch('hilton', "-0.1757", "51.5156", '10000', 'm')
-        env.assertListEqual(res, res2)
+        env.assertEqual(len(res), len(res2))
```
Comment (MeirShpilraien) [inline src/pytest/test.py:968]: Why not sorting the two lists and then compare them? Here you do not compare the content...

File: src/pytest/test.py [hunk 8]
```diff
@@ -961,24 +954,20 @@ def testGeo(env):
...
         res2 = gsearch_inline('hilton', "-0.1757", "51.5156", '10')
-        env.assertListEqual(res, res2)
+        env.assertEqual(len(res), len(res2))
```
Comment (MeirShpilraien) [inline src/pytest/test.py:970]: Same...

File: src/pytest/test.py [hunk 9]
```diff
@@ -992,7 +981,8 @@ def testGeo(env):
         env.assertIn('hotel94', res)
         res2 = gsearch_inline(
             'heathrow', -0.44155, 51.45865, '10', 'km')
-        env.assertListEqual(res, res2)
```
Comment (MeirShpilraien) [inline src/pytest/test.py:995]: Same

File: src/pytest/test.py [hunk 10]
```diff
@@ -1021,22 +1011,29 @@ def testGeoDeletion(env):
...
             'g2', "-0.1757,51.5156",
             't1', "hello")
+    env.cmd('ft.add', 'idx', 'doc3', 1.0, 'fields',
+            'g1', "-0.1757,51.5156",
+            't1', "hello")
 
     # keys are: "geo:idx/g1" and "geo:idx/g2"
-    env.assertEqual(2, env.cmd('zcard', 'geo:idx/g1'))
-    env.assertEqual(2, env.cmd('zcard', 'geo:idx/g2'))
+    env.assertEqual(3, len(env.cmd('FT.DEBUG DUMP_NUMIDX idx g1')[0]))
```
Comment (MeirShpilraien) [inline src/pytest/test.py:1019]: Why it became 3?

File: src/pytest/test.py [hunk 11]
```diff
@@ -1021,22 +1011,29 @@ def testGeoDeletion(env):
             'g1', "-0.1757,51.5156",
             'g2', "-0.1757,51.5156",
             't1', "hello")
+    env.cmd('ft.add', 'idx', 'doc3', 1.0, 'fields',
+            'g1', "-0.1757,51.5156",
+            't1', "hello")
```
Comment (MeirShpilraien) [inline src/pytest/test.py:1016]: Any reason to add another document here? what is it testing that was not tested before?

File: src/spec.c [hunk 12]
```diff
@@ -795,25 +795,23 @@ RedisModuleString *IndexSpec_GetFormattedKey(IndexSpec *sp, const FieldSpec *fs,
...
-      case INDEXFLD_T_GEO:
-        ret = RedisModule_CreateStringPrintf(RSDummyContext, GEOINDEX_KEY_FMT, sp->name, fs->name);
-        break;
       case INDEXFLD_T_FULLTEXT:  // Text fields don't get a per-field index
       default:
         ret = NULL;
         abort();
         break;
     }
+    if (!ret) {
```
Comment (MeirShpilraien) [inline src/spec.c:810]: Can this happened? If yes, do we handle it on all usages? If not lets assert.

File: src/numeric_index.c (hunk 4 above)
Comment (ashtul) [reply src/numeric_index.c:340]: If MIN/MAX are not matched it means the whole block is within range.

File: src/pytest/test.py (hunk 5 above)
Comment (ashtul) [reply src/pytest/test.py:921]: because we are using `strtod` and not rely on GEOADD errors.

File: src/pytest/test.py (hunk 11 above)
Comment (ashtul) [reply src/pytest/test.py:1016]: See below

File: src/pytest/test.py (hunk 10 above)
Comment (ashtul) [reply src/pytest/test.py:1019]: Since we have added above the extra doc.
Thought if we have two indices, we should have different values.

File: src/spec.c (hunk 12 above)
Comment (ashtul) [reply src/spec.c:810]: Not really. put logassert
Should we remove all these functions and use `fs->name` with dict instead?

File: src/numeric_index.c (hunk 4 above)
Comment (MeirShpilraien) [reply src/numeric_index.c:340]: I meant why you check if only if `f->geoFilter == NULL`

File: src/pytest/test.py (hunk 5 above)
Comment (MeirShpilraien) [reply src/pytest/test.py:921]: Why would GEOADD will raise error here and we are not?

File: src/pytest/test.py (hunk 10 above)
Comment (MeirShpilraien) [reply src/pytest/test.py:1019]: :+1:

File: src/spec.c (hunk 12 above)
Comment (MeirShpilraien) [reply src/spec.c:810]: Let's keep it but assert here if it can  not happened

File: src/numeric_index.c (hunk 4 above)
Comment (ashtul) [reply src/numeric_index.c:340]: This is an optimization for numeric which removes the filter in case the whole range is included. It effects `readNumeric` in `inverted_index.c`.
We want to keep this optimization for pure numeric but have to ensure that for geo we do not NULL the `NumericFilter` pointer since we need it for `isWithinRadius` function.

File: src/pytest/test.py (hunk 5 above)
Comment (ashtul) [reply src/pytest/test.py:921]: Before we wouyld pass it to redis as `geoadd idx 1 " 1" doc` and will receive error `ERR value is not a valid float`.
Now that we parse it ourselves, `strtod` just ignores it.

Comment (ashtul) [reply src/pytest/test.py:921]:
```
* Note that this function demands that the string strictly represents
 * a double: no spaces or other characters before or after the string
 * representing the number are accepted. */
int string2d(const char *s, size_t slen, double *dp) {
```
https://github.com/antirez/redis/blob/bdd541174736522c92ff786576e744269c3a1c2c/src/util.c#L497

We are using `strtod` everywhere in the parser so I think it is OK.

File: src/numeric_index.c (hunk 4 above)
Comment (MeirShpilraien) [reply src/numeric_index.c:340]: Ohh I see now, nice :+1:
